Crop Window Toggle

ericmehl opened the pull request:
This changes two related behaviors :
1. Allows all tools, not just non-exclusive tools, to be toggled off with a keyboard shortcut.
2. Adds `Alt+C` shortcut to toggle off the crop window tool as well as the crop window itself, for the plug used for the crop window tool.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on src/GafferUI/View.cpp:
```diff
@@ -268,16 +268,36 @@ void View::toolPlugSet( Gaffer::Plug *plug ) const
+		}
+	}
+
+	bool exclusivePlugActive = false;
 	for( auto &t : Tool::Range( *tools() ) )
 	{
-		if( t != tool && exclusive( t.get() ) )
+		exclusivePlugActive |= ( exclusive( t.get() ) && t->activePlug()->getValue() );
+	}
+	if( !exclusivePlugActive )
```
I think this can just be an `else` block on the `if( tool->activePlug()->getValue() )` conditional can't it? We already know that `tool` is exclusive, so if we know that it has been turned off, then that's all we need isn't it?

johnhaddon replied:
That does make sense - thanks for the clear explanation.

We can change the const-ness of `toolPlugSet()` if we want though - because it's private nobody can be depending on it. What we can't do is add a member, which is unfortunately something we'd want if we were to store a `Signals::Connection` to block, or store a flag to say we're currently calling `setValue()` ourselves.

I does seem like everything would be more predictable if we avoided re-entrancy in `toolPlugSet()` - I'm sure it took some figuring out to determine why the obvious approach didn't work. Since this is all happening on the UI thread, and it's reasonable to assume that it's all local to one View (mucking with the active plugs won't affect another View's plugs), maybe we can just use a static `bool g_ignoreToolPlugSet` that we use to avoid reentrancy? Or we could have that as a member variable and move this to `1.4`. Any preference? Maybe you'd rather stick to what you have since it's tested and working?

ericmehl replied:
I think for clarity I'd prefer to store a `Signals::Connection` so we can block signals when we're doing our default fallback tool selection. Thinking about myself or someone else coming back to this, that sounds like the clearest explanation of what's happening.

It doesn't seem like it's too much of a priority, so pushing it into `1.4` wouldn't be too big of a deal?

If I'm wrong about the priority, the static `bool g_ignoreToolPlugSet` should be fine, I just have a nag in the back of my mind that makes me wary of storing local state in a global variable.

So I'm happy to update an retest, and which solution we go with seems dictated by the priority. Any preference from you on including it in `1.4` vs `1.3`?

johnhaddon replied:
> I think for clarity I'd prefer to store a Signals::Connection so we can block signals when we're doing our default fallback tool selection.

One wrinkle with this would be that we'd need a connection for each tool, which would be a bit more faffy than I made it sound at first. But it's definitely doable.

> I just have a nag in the back of my mind that makes me wary of storing local state in a global variable.

For good reason too! But I can't see a way it would fail in this case.

> Any preference from you on including it in 1.4 vs 1.3?

No huge preference - I agree this isn't a high priority item, so if you'd just like to do the clean version and put it in 1.4 that's fine with me.

ericmehl replied:
Commit https://github.com/GafferHQ/gaffer/pull/5518/commits/34a7926dd0b395ac854135772f5ec5a783c92493 changes the method for activating the default tool to use a `Signal::BlockedConnection` instead of looking to see if all tools were deactivated.

It feels like a pretty good solution...

johnhaddon commented on src/GafferUI/View.cpp:
```diff
@@ -268,16 +268,36 @@ void View::toolPlugSet( Gaffer::Plug *plug ) const
+			if(  order && order->readable() == 0 )
+			{
+				t->activePlug()->setValue( true );
+			}
```
`break`?

ericmehl replied:
Given a `break` in https://github.com/GafferHQ/gaffer/pull/5518/commits/34a7926dd0b395ac854135772f5ec5a783c92493

johnhaddon commented on src/GafferSceneUI/CropWindowTool.cpp:
```diff
@@ -1018,3 +1020,17 @@ Box2f CropWindowTool::resolutionGate() const
+		{
+			const bool newState = !activePlug()->getValue();
+			activePlug()->setValue( newState );
+			m_cropWindowEnabledPlug->setValue( newState );
```
This crashes if `m_cropWindowEnabledPlug` is null, which it can be if the tool hasn't found a plug to edit.

ericmehl replied:
I protected against that crash in https://github.com/GafferHQ/gaffer/pull/5518/commits/3a8c2fdbf2bd4dd0d3a1f34e8f51ec6e5651ac07

ericmehl commented:
I addressed the comments above - despite it being a bit weird that two of the CI builds failed in tests, I think this is ready for another look. Those failures seem unrelated.

johnhaddon commented on src/GafferSceneUI/CropWindowTool.cpp:
```diff
@@ -1018,3 +1020,20 @@ Box2f CropWindowTool::resolutionGate() const
+			activePlug()->setValue( newState );
+			if( m_cropWindowEnabledPlug )
+			{
+				m_cropWindowEnabledPlug->setValue( newState );
```
I think this needs wrapping in an UndoScope.

ericmehl replied:
Good catch! That is added and squash into https://github.com/GafferHQ/gaffer/pull/5518/commits/92729599f03896bb38e0542a68396d737241ddce.

One small, I think, difference - I put the `UndoScope` just above the `m_cropWindowEnabledPlug->setValue()` line. I noticed undo didn't undo the `activePlug()->setValue()`, I think because it's not a descendant of `m_cropWindowEnabledPlug->ancestor<ScriptNode>()`?

I guess it doesn't matter for the functionality, but it seems better homed to put it above  `m_cropWindowEnabledPlug->setValue()` so as not to suggest it will undo `activePlug()->setValue()`.

johnhaddon replied:
Yep, that placement for the UndoScope seems the most logical, and you're right as to why `activePlug()->setValue()` isn't undoable.

johnhaddon reviewed (changes requested):
Sorry Eric, I spotted one more problem that I should have spotted the first time around - see inline comment. Otherwise LGTM - could you squash everything appropriately while adding the UndoScope and then we can merge?

johnhaddon commented:
> Those failures seem unrelated.

Definitely. I've pushed https://github.com/GafferHQ/gaffer/commit/bf248a4d4e44227ba30455fa43fa6c40a08c663b to take care of one of them. The Windows one I've seen before too, but don't know the cause yet.

ericmehl commented:
Squashed and ready for merging as long as my `UndoScope` is in fact placed properly.
